tendermint/src/rpc/event_listener.rs

```diff
@@ -0,0 +1,235 @@
+            EventSubscription::BlockSubscription => "tm.event='NewBlock'",
+        }
+    }
+}
+
+/// Event Listener over websocket.
+/// See: https://docs.tendermint.com/master/rpc/#/Websocket/subscribe
+pub struct EventListener {
+    socket: async_tungstenite::WebSocketStream<TokioAdapter<TcpStream>>,
+}
+
+impl EventListener {
+    /// Constructor for event listener
+    pub async fn connect(address: net::Address) -> Result<EventListener, RPCError> {
+        let (host, port) = match address {
+            net::Address::Tcp { host, port, .. } => (host, port),
+            other => {
+                return Err(RPCError::invalid_params(&format!(
+                    "invalid RPC address: {:?}",
+                    other
+                )));
+            }
+        };
+        //TODO This doesn't have any way to handle a connection over TLS
+        let (ws_stream, _unused_tls_stream) =
+            connect_async(&format!("ws://{}:{}/websocket", host, port)).await?;
+        Ok(EventListener { socket: ws_stream })
+    }
+    //TODO Have a query type instead of a string
+    /// Subscribe to event query stream over the websocket
+    pub async fn subscribe(&mut self, query: EventSubscription) -> Result<(), RPCError> {
+        self.socket
+            .send(Message::text(
+                subscribe::Request::new(query.as_str().to_owned()).into_json(),
+            ))
+            .await?;
+        Ok(())
+    }
+
+    /// Get the next event from the websocket
+    pub async fn get_event(&mut self) -> Result<Event, RPCError> {
+        let msg = self
+            .socket
+            .next()
+            .await
+            .ok_or_else(|| RPCError::websocket_error("web socket closed"))??;
+        match serde_json::from_str::<JsonRPCBlockResult>(&msg.to_string()) {
+            Ok(data) => {
+                if let Some(block_result) = data.0.result {
+                    Ok(Event::JsonRPCBlockResult(Box::new(block_result)))
+                } else {
+                    // The Websocket should never send an empty block
+                    panic!("Websocket sent us an empty block")
+                }
+            }
+            Err(_) => match serde_json::from_str::<JsonRPCTransactionResult>(&msg.to_string()) {
+                Ok(data) => {
+                    if let Some(tx_result) = data.0.result {
+                        Ok(Event::JsonRPCTransactionResult(Box::new(tx_result)))
+                    } else {
+                        // The Websocket should never send an empty transaction
+                        panic!("Websocket sent us an empty transaction")
+                    }
+                }
+                Err(_) => match serde_json::from_str::<serde_json::Value>(&msg.to_string()) {
+                    Ok(json_value) => Ok(Event::GenericJSONEvent(json_value)),
+                    Err(_) => Ok(Event::GenericStringEvent(msg.to_string())),
+                },
```

Comment:
I'm pretty sure this can be simplified in a way that serde returns the correct `Event` enum variant and you don't need matching here again. Non-blocking for merging this PR though.